Add bare metal test for UART.

mprse opened the pull request:
## Description
This is a proposition of UART bare metal test for Green Tea. 

Requirements:

> When vendors start porting a platform to mbed OS, they need to take big steps before they actually can start running tests. One of the first requirements to run greentea is functional UART (RawSerial), unfortunately there's nothing that would tell whether the implementation is correct.
> 
>     Analize greentea to identify serial primitives that need to be tested to ensure the greentea will work correctly.
>     Create tests set:
>         it shouldn't use greentea
>         can be run on minimal (not complete) mbed OS port
>         doesn't use other part of the system
>         tests enough of UART to enable greentea
>         doesnt need to be full unit/integration tests for UART

Commit provides 2 files:
- test.py - host part of the test.  This script builds mbed part of the test and performs serial communication test.
- main.cpp - source file for mbed part of the test. This part tries to communicate with the host test using RawSerial class used by the Green Tea.

Example usage:
- python test.py -t GCC_ARM -m NUCLEO_F070RB -b 9600 - run test using GCC_ARM toolchain, NUCLEO_F070RB board and test serial port with baud rate set to 9600 b/s.

- python test.py -h - see detail usage information.



## Status
HOLD - test location needs to be changed


## Migrations
 NO

mprse commented:
@maciejbocianski @fkjagodzinski - Ready for review.

0xc0170 commented on TESTS/mbed_hal/uart_bare_metal_test/test_config.h:
```diff
@@ -0,0 +1,3 @@
+/* This file has been automatically generated by the python script. */
+/* This file contains serial port baud rate. */
+#define BAUD_RATE 19200
```
We could use config for this? to create Config object, add this attribute there and generate config file that would be ignored by git?

In case, this is the best approach, this file possibly should be created by a  test and being ignored in git? Thus not part of this patch?

mprse replied:
This file is already generated by the script based on provided script parameter. 
I will remove this file from commit.

0xc0170 commented on TESTS/mbed_hal/uart_bare_metal_test/main.cpp:
```diff
@@ -0,0 +1,152 @@
+ * function returns FAILED status.
+ *
+ * */
+test_status test_uart_communication(RawSerial &serial, const char * msg)
```
doxygen comments, in separate header file?

mprse replied:
I'm not sure if we need this for this test. This test must be executed manually and does not use grean tea. The hart of this test is the python script and it provides detailed description.

0xc0170 commented:
retest uvisor

fkjagodzinski commented on TESTS/mbed_hal/uart_bare_metal_test/uart_test.py:
```diff
@@ -0,0 +1,347 @@
+        if('ERROR' in line):
+            build_success = False
+        if('Image:' in line):
+            image_path = line[7:]
```
More stable API sould be used instead of parsing `stdout`. Use `mbed test --compile --build-data foo` and then parse `foo` JSON file:
```python
build_data = {}
with open('foo') as f:
    build_data = json.load(f)

try:
    image_path = build_data['builds'][0]['bin']
except KeyError:
    image_path = ''  # or raise an exception
```

mprse replied:
Fixed as suggested.

fkjagodzinski commented on TESTS/mbed_hal/uart_bare_metal_test/uart_test.py:
```diff
@@ -0,0 +1,347 @@
+        # for like COMnotanumber
+        pass
+    
+    hComPort = win32.CreateFile(port, win32.GENERIC_READ | win32.GENERIC_WRITE, 0, None, win32.OPEN_EXISTING, win32.FILE_ATTRIBUTE_NORMAL | win32.FILE_FLAG_OVERLAPPED, 0)
```
The assumption of using a Windows machine should be documented.

mprse replied:
Note has been added to the script description.

theotherjimmy replied:
You should just not assume that you're on a windows machine! We have linux machines in CI.

mprse replied:
I agree, but currently I don't have linux machine.

mprse commented on TESTS/mbed_hal/uart_bare_metal_test/uart_test.py:
```diff
@@ -0,0 +1,347 @@
+  call host test with appropriate --baudrate parameter.
+  
+  Example usage:
+      uart_test.py -m K64F -m K64F -t GCC_ARM -b 9600
```
Fixed.

mprse commented on TESTS/mbed_hal/uart_bare_metal_test/uart_test.py:
```diff
@@ -0,0 +1,347 @@
+VERBOSE = False
+RETRY_LIMIT = 10
+
+# Listo of expected messages from the mbed device.
```
Fixed.

mprse commented on TESTS/mbed_hal/uart_bare_metal_test/uart_test.py:
```diff
@@ -0,0 +1,347 @@
+    print('[FAIL]: %s') % msg
+    if(serial_port != None and serial_port.isOpen() == True):
+        serial_port.close()
+    sys.exit()
```
Removed additional actions performed by the log function.

mprse commented on TESTS/mbed_hal/uart_bare_metal_test/uart_test.py:
```diff
@@ -0,0 +1,347 @@
+        log_fail('Command can not be executed.')
+        sys.exit()
+    
+    output = str()
```
Fixed.

mprse commented on TESTS/mbed_hal/uart_bare_metal_test/uart_test.py:
```diff
@@ -0,0 +1,347 @@
+        # When dumping output to file both \r and \n will be a new line
+        # To avoid this "extra new-line" we only use \n at the end
+        sys.stdout.write(line.rstrip() + '\n')
+        sys.stdout.flush()
```
Fixed as suggested.

mprse commented on TESTS/mbed_hal/uart_bare_metal_test/uart_test.py:
```diff
@@ -0,0 +1,347 @@
+                'PASSED\0']
+
+def log_info(msg):
+    print('[INFO]: %s') % msg
```
Thanks for the tip, but for now I decided to stay with my log functions. After the update they are only simple wrappers to print function. This gives additional flexibility. I can have log functions to indicate failure, pass and received/transmitted data.

theotherjimmy replied:
And how is this an advantage over the logging module?

mprse replied:
This solution gives me everything what is needed to print out logs and allows to format log messages. Of course there is no problem to use logging module instead.

bridadan replied:
I'd recommend moving to logging as well. I know the tools don't currently support Python 3, but in the future we'll have to (Python 2 is only minimally being maintained). logging has Python 3 support out of the box, where as this current implementation does not.

mprse replied:
Modified logging method as suggested.

0xc0170 commented:
@mprse jenkins CI fails, related to this patch, please check

mprse commented:
@0xc0170 

> @mprse jenkins CI fails, related to this patch, please check

I'm not sure if there is a sens to check this tests on build machines. It needs to be executed manually. 
Jenkins fails with the following error:

> Error:  #5: cannot open source input file "test_config.h": No such file or directory

test_config.h is a config file generated by python script (has been removed from this patch).

mprse commented:
Modified function to reset the mbed board. Current implementation should be valid for Windows and Linux (not tested).

0xc0170 commented:
> test_config.h is a config file generated by python script (has been removed from this patch).

I see,  than this test should not be automatically run, as it has this python dependency. is it required or how to make it CI friendly?

mprse commented:
> I see, than this test should not be automatically run, as it has this python dependency. is it required or how to make it CI friendly?

Honestly I'm not familiar with CI, maybe we should put it in some place where CI is not performed.

This test is for vendors to be able to verify if their serial HAL drivers are ready to use Grean Tea. To do this python script should be run with some parameters. This script builds the embed app, loads it on the board and run it. This embed app uses only RawSerial class and tries to communicate with PC (python script) by sending sample strings with different length and waiting for the responses. If all strings are successfully transmitted in both sides, then test is passed.

0xc0170 commented:
> Honestly I'm not familiar with CI, maybe we should put it in some place where CI is not performed.

cc @studavekar @theotherjimmy - how to handle this type of test?

studavekar commented:
We can add in check, if and only if there is a label "needs: CI" then we start CI.

0xc0170 commented:
> This test is for vendors to be able to verify if their serial HAL drivers are ready to use Grean Tea. To do this python script should be run with some parameters. This script builds the embed app, loads it on the board and run it. This embed app uses only RawSerial class and tries to communicate with PC (python script) by sending sample strings with different length and waiting for the responses. If all strings are successfully transmitted in both sides, then test is passed.

@studavekar @bridadan How we can handle this one, via ignore or any other suggestions?

bridadan commented on TESTS/mbed_hal/uart_bare_metal_test/uart_test.py:
```diff
@@ -0,0 +1,342 @@
+    required = parser.add_argument_group('required arguments')
+    optional = parser.add_argument_group('optional arguments')
+    required.add_argument("-m", "--target", help="Target id")
+    optional.add_argument("-t", "--toolchanin", help="Compile toolchain. Example: ARM, GCC_ARM, IAR.", default='GCC_ARM')
```
Typo here, should be `--toolchain`

mprse replied:
Fixed.

bridadan commented:
> @studavekar @bridadan How we can handle this one, via ignore or any other suggestions?

One option would be to create a new folder in mbed OS for these kinds of tests. This one would be included in a `.mbedignore`.

To actually build the test, you could force tools to ignore the `.mbedignore` by using the `--source` option:

```
/mbed-os
   - /bare_metal_tests
       - /.mbedignore (contents would be just "*" with no quotes)
       - /uart
             - /main.cpp
             - /test.py
```

Where test.py would call mbed-cli as:
```
mbed test --source <path to mbed-os root> --source <path to mbed-os root>/bare_metal_tests/uart
```

mprse commented:
@bridadan I tried to use "mbed test --source" in current tree folder where test is located in:
_/m-bed/TESTS/mbed_hal/uart_bare_metal_test/main.cpp_

I have executed the following command:
`mbed test --source . --source TESTS/mbed_hal/uart_bare_metal_test -m K64F -t GCC_ARM -c`
The compilation process have failed. It looks like all test source files have been included into this build since I get _multiple definition of `main'_ error messages.

Currently in python script I use _mbed --compile_ to build test and I have my functions for flashing and resetting the board. As far as I understand these function will be redundant. Command
`mbed test --source . --source TESTS/mbed_hal/uart_bare_metal_test -m K64F -t GCC_ARM -c`
will flash and reset the board for me. Correct?

@0xc0170
Shell I modify this pull request and update the directory structure to by consistent with proposition provided by @bridadan?

theotherjimmy commented:
@0xc0170 @bridadan @studavekar Could you review?

bridadan commented:
> The compilation process have failed. It looks like all test source files have been included into this build since I get multiple definition of `main' error messages.

Darn it, you're right. I forgot that that's how that particular script works.

I did some more digging and learned that I goofed when I originally wrote the file. It doesn't actually use the `--source` flag when it's searching for tests, instead it uses your current working directory or the `-p` flag, which I don't think anything _really_ uses. This is a problem for this kind of scenario.

SO I patched it: https://github.com/bridadan/mbed/commit/f8ed50794eb6b4941b3f2f34cfc509f22f50d054

You'll also need to update your directory structure:

```
/mbed-os
- /.mbedignore (contents would be just "*" with no quotes)
- /bare_metal_tests
    - /TESTS
        - /bringup
            - /uart
                - /main.cpp
                - /test.py
```

And then the CLI command would look like:

```
cd mbed-os
mbed test -m K64F -t GCC_ARM --source . --source bare_metal_test
```

I have to say, I'm not really seeing that many advantages this has over just using Greentea/htrun. It basically invokes the build tools, copies the binaries, resets the target, and compare strings in the Python script. Which basically describes Greentea in a nutshell.

It looks like you're citing a engineering requirement @mprse, can you tell us where this is from?

mprse commented:
@bridadan Thank you for clarifications.

The cited requirement can be found here:
https://jira.arm.com/browse/IOTMORF-1244

I have updated directory tree and with your patch to test.py I was able to run the mbed part of the test with:
`mbed test -m K64F -t GCC_ARM --source . --source bare_metal_test` command. As we expect this command successfully built the embed test, loaded the binary file on the device and reset the device. Unfortunately after this operation script sends synchronisation packets to the device and gets no response. This results in synchronisation error after a while. For me the best option would be to end the `mbed test` command after the reset of the board, but it looks like it is not possible. For now I have updated build command in python part of the test to the following form : `mbed test -m K64F -t GCC_ARM --source . --source bare_metal_test --compile` and left my functions to flash and reset the target.

fkjagodzinski commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,331 @@
+import mbed_lstools
+from subprocess import Popen, PIPE, STDOUT
+import shutil
+from serial import win32
```
Unused import statement.

mprse replied:
Removed.

fkjagodzinski commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,331 @@
+    @param cmd command to be executed.
+    @return image path.
+    """
+    build_success = True
```
Unused var.

mprse replied:
Removed.

fkjagodzinski commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,331 @@
+       a break signal to the device.
+       
+    @param port serial port id.
+    @param delay_break break signal time [s].
```
Unused arg.

mprse replied:
Removed.

fkjagodzinski commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,331 @@
+ 
+    break_port = serial.Serial(port=port, baudrate=9600)
+    
+    break_port.sendBreak()
```
Is this the place for `delay_break`, like `break_port.sendBreak(delay_break)` ?

mprse replied:
delay_break variable has been removed.

bridadan commented:
@mprse You can stop `mbed test` from flashing/testing the board by adding the `--compile` flag to the command.

theotherjimmy commented on .mbedignore:
```diff
@@ -0,0 +1 @@
+*
```
This should probably not be in the root, and was meant for `bare_metal_tests`

mprse replied:
Content of the .mbedignore file has been modified to "bare_metal_tests/*".
This should allow to run bear metal tests and also regular tests.

bridadan commented:
I had an offline conversation with @bulislaw and @theotherjimmy. We wanted to make sure that these changes don't start adding the same features that Greentea has, thus defeating the purpose of the changes.

So we came away with the following suggestions:

1. Build the tests with `mbed compile` and tell the implementer to flash them however appropriate for their board
    1. Greentea handles the automatic copying of boards, so we should stay away from doing this in these scripts to keep the process simple and easy to understand
    2. This will stil use the `--source` options ex. `mbed compile --source . --source <path to test>`
2. Have the implementer reset their board however appropriate for their board
3. The implementer should then start the Python script with the correct COM port/serial device
    1. This script should just use pyserial to open a com port and check that data is transferring correctly

Hopefully that all makes sense, let me or @bulislaw know if you have any questions. We can also take this discussion to the JIRA ticket if need be.

As for specifics, I think using the .mbedignore with the new directory structure still works. But you can simplify the structure now since this won't be inside the normal "TESTS" directories.
So something like `/mbed-os/bare_metal_tests/uart/<all test files>`. This would correspond to a command of `mbed compile --source . --source bare_metal_tests/uart`.

bulislaw commented:
We will create section in porting guide covering details of how to run the low level/bare metal tests. They will be only useful on the beginning of porting process so it can be more manual.

mprse commented:
@bridadan @bulislaw @theotherjimmy 
Thanks for the detail requirements for the UART test.
I have added second commit which implements UART test based on provided suggestion.

Directory tree:
```
/mbed-os
- /.mbedignore (content: "bare_metal_tests/*" with no quotes)
- /bare_metal_tests
  - /uart
    - /main.cpp
    - /test.py
```
This allows to build UART test with the following command (assuming that we are in `mbed-os` dir):
`mbed compile --source bare_metal_tests/uart --source .`

Tested also that regular tests can be successfully run using `mbed test` command.

Steps to execute the test:
- user runs the python script (test.py) providing at least the target name (e.g. `test.py -t GCC_ARM -m NUCLEO_F070RB -b 9600`).
- script builds the mbed test using the following command:
`mbed compile -t <TOOLCHAIN> -m <TARGET> --source . --source ../.. --build-data build.log [-c]`
- script requests user to flash and reset the mbed board and confirm that board is ready for the test by pressing ENTER,
- script performs UART communication test,
- script reports the test status.

Note:
Build log is used to determine if build process has been successfully performed.

> We can also take this discussion to the JIRA ticket if need be.

I'm not sure if this is consistent with the work flow, but I added link to this PR in the JIRA task, so the progress can be seen here.

> We will create section in porting guide covering details of how to run the low level/bare metal tests. 
They will be only useful on the beginning of porting process so it can be more manual.

Shell I add some readme file or everything will be in mentioned manual?

theotherjimmy commented on .mbedignore:
```diff
@@ -0,0 +1 @@
+bare_metal_tests/*
```
Instead of placing this in the root with `bare_metal_tests/*` could you place it in `bare_metal_tests` with `*` please? I don't want to set a precedent of having root level `.mbedignore` files just yet.

mprse replied:
Fixed directory tree as suggested. Verified that bare-metal test and regular tests can be successfully executed.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+    # Get image path     
+    try:
+        with open('build.log') as f:
+            print('file ok')
```
This should be logging. I would not expect a user to need this information.

mprse replied:
Sorry for that. These are left debug prints which are not present in the second/final version.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+        with open('build.log') as f:
+            print('file ok')
+            build_data = json.load(f)
+            print('json ok')
```
This print should also be logging.

mprse replied:
Sorry for that. These are left debug prints which are not present in the second/final version.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+            build_data = json.load(f)
+            print('json ok')
+            image_path = build_data['builds'][0]['bin']
+            print('json ok')
```
This print should be logging, and is the same as the prior print. Maybe you meant `image found` or similar?

mprse replied:
Sorry for that. These are left debug prints which are not present in the second/final version.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+            print('json ok')
+            image_path = build_data['builds'][0]['bin']
+            print('json ok')
+    except:
```
Could you narrow the exceptions you catch here? As it stands it would catch things like `C-c` and out of memory conditions.

mprse replied:
Fixed.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+ 
+    break_port = serial.Serial(port=port, baudrate=9600)
+    
+    break_port.sendBreak()
```
Do all boards reset when a serial break is sent? is that something we can rely on?

mprse replied:
The new version assumes that user performs board reset manually. Reset function does not exist in the new version.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+    logging.info('source: %s' % (src))
+    logging.info('destination: %s' % (dst))
+    
+    shutil.copyfile(src, dst)
```
Can we reliably flash a new target by copying in this manner?

mprse replied:
The new version assumes that user performs board flashing manually. Flashing function does not exist in the new version.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+        baudrate=baudrate,
+        parity=serial.PARITY_NONE,
+        stopbits=serial.STOPBITS_ONE,
+        bytesize=serial.EIGHTBITS)
```
The indention of this line, and the prior 4 lines is incorrect. Could you fix it?

mprse replied:
Fixed.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+        line += c
+        if(c == '\0'):
+            logging.debug('[RECV] %s' % line)
+            return line
```
This line, and the prior return look to be the same. Could you de-duplicate this code? Further, the `if ... continue` is the same as an `if ... else`. I think that an `if  ... else` would be more clear here.
For example:
```python
    while(True):
        if(serial_port.in_waiting == 0):
            time.sleep(0.001)
            limit += 1
            if (limit == RETRY_LIMIT):
                break
        else:
            limit = 0
            c = serial_port.read()
            line += c
            if(c == '\0'):
                break
    logging.debug('[RECV] %s' % line)
    return line

```

mprse replied:
Fixed as suggested.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+            if not line.endswith('\0'):
+                line += '\0'
+            serial_write_line(line)
+        time.sleep(1)
```
Is this sleep needed? your `serial_read_line` function already handles delays in serial reception.

mprse replied:
Sleep is not needed and has been removed.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+    
+    # Prepare required/optional arguments.
+    parser = argparse.ArgumentParser(description='Execute UART bare metal test to check if platform is ready to use green tea.')
+    parser._action_groups.pop()
```
This is not a documented part of the argparse api (it starts with an `_`). Please don't use it.

mprse replied:
Modified argument parsing.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+    parser._action_groups.pop()
+    required = parser.add_argument_group('required arguments')
+    optional = parser.add_argument_group('optional arguments')
+    required.add_argument("-m", "--target", help="Target id")
```
If this argument is required, you should add a `required=True` parameter.

mprse replied:
Fixed as suggested.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+        if(mut['platform_name'] == args.target):
+            sel_platform_id = i
+            sel_str = '  --- SELECTED ---'
+        print '%-20s %-10s %-10s %-10s' % (mut['platform_name'], mut['serial_port'], mut['mount_point'], sel_str)
```
This print should be a logging statement.

mprse replied:
In the second version all prints were modified to logging.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+        sel_str = ''
+        if(mut['platform_name'] == args.target):
+            sel_platform_id = i
+            sel_str = '  --- SELECTED ---'
```
Why is this so complicated; why can't this `break`?

mprse replied:
This is done to print list of all available platforms and store the id of the selected platform for further processing. `break` can to be used since we want to show all platforms.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+from subprocess import Popen, PIPE, STDOUT
+import shutil
+import argparse
+import ntpath
```
Do not do this. It's not cross platform AT ALL. Use `os.path` instead.

mprse replied:
Removed redundant imports (including ntpath).

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+    file = open('test_config.h','w') 
+    file.write('/* This file has been automatically generated by the python script. */\n')
+    file.write('/* This file contains serial port baud rate. */\n')
+    file.write('#define BAUD_RATE %s\n' % baudrate)
```
This could be done by `mbed compile ... -DBAUD_RATE=%s`.

mprse replied:
That is much better. Fixed as suggested.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+    # Get image path     
+    try:
+        with open('build.log') as f:
+            print('file ok')
```
Should be logging.

mprse replied:
Sorry for that. These are left debug prints which are not present in the second/final version.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+        with open('build.log') as f:
+            print('file ok')
+            build_data = json.load(f)
+            print('json ok')
```
Should be logging.

mprse replied:
Sorry for that. These are left debug prints which are not present in the second/final version.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+            build_data = json.load(f)
+            print('json ok')
+            image_path = build_data['builds'][0]['bin']
+            print('json ok')
```
Should be logging, and it's the same as the print above. Maybe you meant something similar to `image path ok`?

mprse replied:
Sorry for that. These are left debug prints which are not present in the second/final version.

theotherjimmy commented on bare_metal_tests/TESTS/bringup/uart/test.py:
```diff
@@ -0,0 +1,330 @@
+    if(clean_flag):
+        clean = '-c'
+    
+    cmd = 'mbed test -t %s -m %s --source . --source bare_metal_tests --build-data bare_metal_tests/TESTS/bringup/uart/build.log --compile %s' % (toolchain, target, clean)
```
Could you split this line at 80 characters? It's overrunning the line as it stands.

mprse replied:
Fixed.

theotherjimmy commented on bare_metal_tests/uart/test.py:
```diff
@@ -0,0 +1,300 @@
+from subprocess import Popen, PIPE, STDOUT
+import shutil
+import argparse
+import ntpath
```
Same here. Use `os.path` instead.

mprse replied:
Already fixed.

mprse commented:
Made one final commit from the existing two commits and modified directory tree after review:

```
/mbed-os
- /bare_metal_tests
  - /.mbedignore (content: "*" with no quotes)
  - /uart
    - /main.cpp
    - /test.py
```

0xc0170 commented:
Thanks @mprse, having this first serial tests is beneficial. After these changes, is it actually the proper location this repository and not being part of own module that can be linked in the porting documentation ?

bulislaw commented:
Let's keep things simple and in one place.

theotherjimmy commented:
Thanks! I'll review this again now.

theotherjimmy commented on bare_metal_tests/uart/main.cpp:
```diff
@@ -0,0 +1,151 @@
+
+int main()
+{
+    RawSerial serial(USBTX, USBRX, BAUD_RATE);
```
Where does `BAUD_RATE` come from? Could it be moved to a configuration file?

mprse replied:
BAUD_RATE symbol is provided via `mbed compile -D BAUD_RATE` as you suggested.
Configuration file generated by the python script has been replaced by suggested compiler option.

theotherjimmy commented on bare_metal_tests/uart/main.cpp:
```diff
@@ -0,0 +1,151 @@
+        }
+    }
+
+    /* Just wait. */
```
I don't think this comment is beneficial. I'm pretty sure that most people that know some C can tell that `while (true)` means spin forever.

mprse replied:
Comment has been removed.

theotherjimmy commented on bare_metal_tests/uart/test.py:
```diff
@@ -0,0 +1,282 @@
+
+    # Check process status.
+    returncode = p.wait()
+    if returncode < 0:
```
This should be:
```python
if returncode != 0
```
to catch positive return codes too.

mprse replied:
Fixed.

theotherjimmy commented on bare_metal_tests/uart/test.py:
```diff
@@ -0,0 +1,282 @@
+    try:
+        p = Popen(cmd, stdout=sys.stdout)
+    except OSError:
+        logging.error('Command can not be executed.')
```
Please use `logging.exception` here instead.

mprse replied:
Used logging.exception in `except` section.

theotherjimmy commented on bare_metal_tests/uart/test.py:
```diff
@@ -0,0 +1,282 @@
+                                    stopbits=serial.STOPBITS_ONE,
+                                    bytesize=serial.EIGHTBITS)
+
+    except:
```
Could you narrow what this catches? At the moment it would catch many many errors that you probably don't intend to catch.

mprse replied:
Specified exceptions to catch.

theotherjimmy commented on bare_metal_tests/uart/test.py:
```diff
@@ -0,0 +1,282 @@
+    @param line line to be written.
+    """
+    serial_port.write(line)
+    logging.debug('[SEDN] %s' % line)
```
This looks like a typo `SEDN` -> `SEND`

mprse replied:
Fixed.

theotherjimmy commented on bare_metal_tests/uart/test.py:
```diff
@@ -0,0 +1,282 @@
+  
+    # Perform serial comunication test with the mbed device.
+    for msg in message_list:
+        if (wait_for_message_and_respond(msg) == False):
```
Please use:
```python
if not wait_for_message_and_respond(msg):
```
it's more pythonic.

mprse replied:
Fixed.

mprse commented:
> I have a general improvement request for the script: Could we avoid global variables? If you want suggestions on how to eliminate the use of global variables, let me know.

Most global variables have been removed (only RETRY_LIMIT symbol is now global).
I'm not sure if provided solution meets your expectations. If no, then please provide your suggestion.

theotherjimmy commented:
Thanks. I'll be able to review this again in a few hours. I was mostly
concerned with the serial global variable.

On Oct 27, 2017 8:10 AM, "Przemek" <notifications@github.com> wrote:

> I have a general improvement request for the script: Could we avoid global
> variables? If you want suggestions on how to eliminate the use of global
> variables, let me know.
>
> Most global variables have been removed (only RETRY_LIMIT symbol is now
> global).
> I'm not sure if provided solution meets your expectations. If no, then
> please provide your suggestion.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ARMmbed/mbed-os/pull/5273#issuecomment-339966821>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAp1YT0CBKXYXOGVr7W2pvB6Z1XHbo_lks5swdYugaJpZM4PyEG2>
> .
>

theotherjimmy commented on bare_metal_tests/uart/test.py:
```diff
@@ -0,0 +1,276 @@
+    for msg in message_list:
+        if not wait_for_message_and_respond(serial_port, msg):
+            logging.error("Expected message has NOT been received by the host.")
+            sys.exit();
```
Extraneous semicolon (`;`)

mprse replied:
Fixed.

theotherjimmy commented on bare_metal_tests/uart/test.py:
```diff
@@ -0,0 +1,276 @@
+    # List of expected messages from the mbed device.
+    message_list = ['x\0', 
+                    'mbed\0',
+                    'Green Tea UART bear metal test.\0', 
```
`bear` -> `bare` ?

mprse replied:
Fixed.

bulislaw commented:
@theotherjimmy are all your comments answered now?

theotherjimmy commented:
@mprse @bulislaw How do you propose that we test this in CI?

bridadan commented:
@theotherjimmy Since the whole point of this test is to enable automated hardware testing, I'd suggest only doing the compile step in Travis or Jenkins.

mprse commented:
@theotherjimmy I'm not familiar with CI testing, but it looks like we could only check if mbed part of the test can by successfully build using the following command: 
`mbed compile -t <TOOL_CHAIN> -m <TARGET> --source . --source ../.. -c --build-data build.log -D BAUD_RATE=<BAUD_RATE>`

`BAUD_RATE` symbol must be defined.

bulislaw commented:
I would like to have it tested regularly to make sure it's up to date with the APIs.

theotherjimmy commented:
> the test can by successfully build using the following command:
>```bash
>mbed compile -t <TOOL_CHAIN> -m <TARGET> --source . --source ../.. -c --build-data build.log -D BAUD_RATE=<BAUD_RATE>
>```
You assume that mbed compile would be run in `bare_metal_tests/uart`, which involves changing directory in the CI. I advise against this, especially because the following command, executed from the project root, does the same thing.
```bash
mbed compile -t gcc_arm -m k64f --source . --source bare_metal_tests/uart -c --build-data build.log -D BAUD_RATE=9600
```

bulislaw commented:
@mprse ping

mprse commented:
@bridadan @theotherjimmy @0xc0170 
I'm don't have experience with CI and Travis. Can you provide some information what is the proper way to add compilation of this test to Travis. There is a `.travis.yml` which contains build script. As far as I understand compile command needs to be added here. Additionally which tool-chains should be used? If we want only to check if RawSerial API is up to date, then we can use only one e.g. GCC_ARM.

mprse commented:
Added uart bare metal test to Travis, but I'm not sure if this is the correct way to do that. Unfortunately now we get timeout error, can we increase the time limit for Travis?

> The job exceeded the maximum time limit for jobs, and has been terminated.

0xc0170 commented:
We have noticed this recently, we should log this as an issue and provide some info how to address this

bulislaw commented:
Is there something wrong with the CIs? the test seem to be hanging or what? @mprse @0xc0170

0xc0170 commented:
/morph build

0xc0170 commented:
Restarted uvisor. @alzix To be certain, the log is java exception, but is it related to actual test ? Please review build 581

alzix commented:
@0xc0170 - the log seems to clean now

0xc0170 commented:
@fkjagodzinski Can you review please the updates?

fkjagodzinski commented on bare_metal_tests/uart/test.py:
```diff
@@ -0,0 +1,276 @@
+          '--build-data build.log %s -D BAUD_RATE=%s' \
+          % (toolchain, target, clean, baudrate)
+    
+    return run_command(cmd)
```
I can't see any benefit of `run_command` being a separate function. This might as well continue here, in `build_test`.

0xc0170 commented:
The last bits.

This might be valuable test, but why is it in the root of mbed-os? `bare_metal_tests/uart/main.cpp` does not look like a correct placement. I asked previously for the better placement. I would imagine having this in the porting documentation (thus having a separete repo would not be an issue, if this does not fit here). I do not like having this as it is now. It can't go into TESTS, not in tools, so if there is no correct placement, create a new repo and docs will point to that. I do not see a problem there. Travis config in this repo can contain a test config for that. 

Note, regarding adding tests to travis, you add them as mbed 2. This works, however not what we want to achieve. Can not this be added as a configuration to travis. There's an update in  travis in littlefs PR that lands soon, it will update travis to have test units (groups in travis, travis would show mbed-2 tests, littlefs test,etc, not a huge log as it is now). I would imagine this there as a new test unit, as it's not mbed 2 test. This is just a detail, I would like to find a proper placement for this first.

mprse commented:
Directory structure has been proposed by @bridadan and I thought that this is the final location. Maybe the valid location would be `features` directory?

adbridge commented:
I agree with @0xc0170 I think a suite of tests that are geared purely for the porting guide would benefit from being in their repo. Mbed-os clone is big enough already without cluttering with additional tests that are not directly related to the codebase...

bulislaw commented:
From what I rememember @bridadan and @theotherjimmy suggested to put it in TEST directory and add mbedignore so that it's ignored by mbed test command.

mprse commented:
@0xc0170 @bulislaw @acabarbaye 

> It can't go into TESTS, not in tools..

> From what I rememember @bridadan and @theotherjimmy suggested to put it in TEST directory and add mbedignore so that it's ignored by mbed test command.

Do we have some decision about the test location? 
If TEST directory can be used it looks like it is the most suitable place.

bulislaw commented:
@bridadan @theotherjimmy is there anything stoping us from having it in `TEST/`?

bridadan commented:
@bulislaw 

> From what I rememember @bridadan and @theotherjimmy suggested to put it in TEST directory and add mbedignore so that it's ignored by mbed test command.

Pretty sure that should be fine!

theotherjimmy commented:
@bulislaw That should be fine.

sg- commented:
Why doesn't the test case use unity?

bridadan commented:
@sg- Unity relies on utest for printing: https://github.com/ARMmbed/mbed-os/blob/master/features/frameworks/unity/unity/unity_config.h#L38

And utest relies on Greentea for printing: https://github.com/ARMmbed/mbed-os/blob/729ef153076ea9db4a9c9f81e90076960e035213/features/frameworks/utest/source/unity_handler.cpp#L38

And the whole point of this test is to not use Greentea.

sg- commented:
> And the whole point of this test is to not use Greentea.

So if `void utest_safe_putc(int chr)` was WEAK that would make the whole thing fit? I could imagine wanting to use SWO or a DCC like system to put and get data from a physical transport.

bridadan commented:
@sg- after looking at this a bit more, I'm now remembering that you **can't use Unity outside of utest**. This is due to the way Unity allows you to jump out of a utest test case immediately as soon as an assert fails (known as "Fail and bail": https://github.com/ARMmbed/mbed-os/blob/master/features/frameworks/unity/source/unity.c#L20).

Unity asserts only work in the context of a utest test case. So if you wanted to use Unity asserts in non-utest test cases, some further refactoring would need to take place.

sg- commented:
In that case, is this essentially a greentea self-test?

bulislaw commented:
I wouldn't say so as it doesn't test greentea, it validates that your serial driver is good enough to run further tests (which use greentea).

mprse commented:
Do we have final decision about the localisation of this test?

0xc0170 commented:
> @bridadan @theotherjimmy is there anything stoping us from having it in TEST/?

This got voted, can we place it there

adbridge commented:
@mprse Looks like this now needs a rebase ? 
@sg- Do you have an opinion on whether the bare metal tests should actually live within mbed-os or have their own repo? Considering they are effectively just testing that parts of the hardware themselves function sufficiently to actually run mbed tests on them ?

mprse commented:
> @mprse Looks like this now needs a rebase ?

Fixed conflicts - updated Travis script.

mprse commented:
@0xc0170 @sg @bulislaw @bridadan @theotherjimmy @adbridge - can we make a final decision here about the test location?

adbridge commented:
I still maintain these should go to their own repository. The logic behind this is quite clear, they are an alternative to Greentea to ascertain whether or not the drivers are in a suitable state to be able to work with Greentea. Greentea has it's own repository so why shouldn't this ?

cmonr commented:
Will need to be put into another repo.
